### DISCUSSION TITLE
How to override templates

### DISCUSSION THREAD

**Discussion by ben519** [squidfunk/mkdocs-material#3406] [Show and tell]
Recently solved a problem by overriding a template. Thought I'd document my steps as a mini tutorial for others.

## Problem
I plan on having a lot of top-level sections in my `nav` like 

```
theme:
  name: material
  features:
    - toc.integrate
    - navigation.tabs 

nav:
  - Home: index.md
  - Contributors: contributors.md
  - Super Awesome Course 1:
    - courses/course1/index.md
    - Section 1:
      - courses/course1/section1/example1.md
    - Section 2:
      - courses/course1/section2/example2.md
      - courses/course1/section2/example3.md
  - Super Awesome Course 2: course2.md
  - Super Awesome Course 3: course3.md
  - Super Awesome Course 4: course4.md
  - Super Awesome Course 5: course5.md
  - Super Awesome Course 6: course6.md
```

The problem is, this creates a very messy header..

<img width="1232" alt="Screen Shot 2022-01-01 at 2 16 28 PM" src="https://user-images.githubusercontent.com/723099/147860868-88e358cf-bf7e-408b-b6df-485b7ee063e0.png">

So, I'd like to exclude *Super Awesome Course 1*, *Super Awesome Course 2*, ... from the header. (They'll still be accessible via links on my home page.) **How can I exclude these links from the header?**

## Solution
A good solution is to tweak the html template to exclude them. The final html for the site is generated from a spider web of *templates* that reference variables and other templates. The whole thing is based on a templating engine called [Jinja](https://jinja.palletsprojects.com/en/3.0.x/).

### Which Template?
First question is **which template do I override?**.

To answer this, inspect the header in chrome developer tools and look for a unique element identifier for tabs..

<img width="820" alt="Screen Shot 2022-01-01 at 4 00 16 PM" src="https://user-images.githubusercontent.com/723099/147861037-df0aa569-7e17-44c8-871f-6899c53e4e37.png">

"md-tabs__list" looks pretty identifying. Let's search for it in the repo..

<img width="1395" alt="Screen Shot 2022-01-01 at 4 01 43 PM" src="https://user-images.githubusercontent.com/723099/147861055-0b322a62-aa1d-4f64-9293-60d74cf006f6.png">

[`src/partials/tabs.html`](https://github.com/squidfunk/mkdocs-material/blob/b3c0163b3b6b6e4d2cb4d50fbd535018042c4970/src/partials/tabs.html) looks like what we want. 

### How to override?
The code here is quite simple.

```
<!-- Navigation tabs -->
<nav
  class="md-tabs"
  aria-label="{{ lang.t('tabs.title') }}"
  data-md-component="tabs"
>
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      {% for nav_item in nav %}
        {% include "partials/tabs-item.html" %}
      {% endfor %}
    </ul>
  </div>
</nav>
```

Our goal is to modify it like this, so that only *Home* and *Contributors* show in the header.

```
<!-- Navigation tabs -->
<nav
  class="md-tabs"
  aria-label="{{ lang.t('tabs.title') }}"
  data-md-component="tabs"
>
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      {% for nav_item in nav %}

        {% if nav_item in ['Home', 'Contributors'] %}
            {% include "partials/tabs-item.html" %}
        {% endif %}

      {% endfor %}
    </ul>
  </div>
</nav>
```

Here's how we do it (following the [instructions here](https://squidfunk.github.io/mkdocs-material/customization/#extending-the-theme)).

1. Copy [`src/partials/tabs.html`](https://github.com/squidfunk/mkdocs-material/blob/b3c0163b3b6b6e4d2cb4d50fbd535018042c4970/src/partials/tabs.html) to `overrides/partials/tabs.html` where `overrides/` is a sibling directory of `docs/`
2. Tell MkDocs you want it to use your `overrides/` directory by setting the `custom_dir` attribute in `mkdocs.yml` like

```
theme:
  name: material
  custom_dir: overrides
```

3. Modify the code in `tabs.html` to make use of the if statement (as shown above ^).
4. `mkdocs serve` as usual to build the site..

**This doesn't work.**

### How to debug?
To see what went wrong, change `tabs.html` like so

```
<!-- Hack: unset variable, as we're using it recursively in tabs-item.html -->
{% set class = "" %}

<!-- Navigation tabs -->
<nav
  class="md-tabs"
  aria-label="{{ lang.t('tabs.title') }}"
  data-md-component="tabs"
>
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      {% for nav_item in nav %}

        <h1>{{nav_item}}</h1>

      {% endfor %}
    </ul>
  </div>
</nav>
```
and re-serve the site.

<img width="932" alt="Screen Shot 2022-01-01 at 4 15 50 PM" src="https://user-images.githubusercontent.com/723099/147861258-ba2e4541-a798-498b-ae29-35eb4b9cc72f.png">

As you can see, `nav_item` is an object. To get the *title*, we need to use `nav_item.title`. Again, we change our overridden tabs.html file as

```
<!-- Hack: unset variable, as we're using it recursively in tabs-item.html -->
{% set class = "" %}

<!-- Navigation tabs -->
<nav
  class="md-tabs"
  aria-label="{{ lang.t('tabs.title') }}"
  data-md-component="tabs"
>
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      {% for nav_item in nav %}

        {% if nav_item.title in ['Home', 'Contributors'] %}
            {% include "partials/tabs-item.html" %}
        {% endif %}

      {% endfor %}
    </ul>
  </div>
</nav>
```

Alas, it works!

<img width="347" alt="Screen Shot 2022-01-01 at 4 19 31 PM" src="https://user-images.githubusercontent.com/723099/147861314-8b2c79cc-d937-4a77-82d6-9d483cdca241.png">

### How to use a variable
One more thing.. I don't want to tweak `tags.html` every time I want to include or exclude a page. It'd be nicer if I could create a variable in `mkdocs.yml` like `include_headers = ['Home', 'Contributors']` which I could reference in the `tabs.html` template.

To set this up, create the `header_items` variable as an attribute below `extra`. Like this

```
extra:
  generator: false
  header_items:
    - Home
    - Contributors
```

Then in `tabs.html` change

```
{% if nav_item.title in ['Home', 'Contributors'] %}
```

to

```
{% if nav_item.title in config.extra.header_items %}
```

**Comment by squidfunk** [comment]
Thanks! Great tutorial!

**Comment by SPyofgame200** [comment]
While override is recommended, I found a solution for certain things that don't have their partials HTML parts


![image](https://github.com/squidfunk/mkdocs-material/assets/52689671/b317fc0d-5c17-4096-a8bb-f27c45e9567a)

By using js injection in extra_Js

```js
const logoContainer = document.querySelector('.md-header__button.md-logo');
const logoImage = logoContainer.querySelector('img');
const iconifyStructure = `
<div class="iconify">
 <div class="iconify_wrapper">
    <div class="iconify_box">
      <div class="iconify_box-inner">
        <div class="iconify_bar"></div>
        <div class="iconify_bar"></div>
        <div class="iconify_bar"></div>
        <div class="iconify_bar"></div>
        <img class="iconify_img" alt="${logoImage.alt}" src="${logoImage.src}">
      </div>
    </div>
 </div>
</div>
`;
logoContainer.innerHTML = iconifyStructure;
```

**Reply by kamilkrzyskow** [reply]
GitHub search isn't the best also you've searched including the dots which aren't in the HTML

https://github.com/squidfunk/mkdocs-material/blob/4eb1a4393741b1ecf8019dd53292b7a4d02e16f5/src/templates/partials/header.html#L42

https://github.com/squidfunk/mkdocs-material/blob/4eb1a4393741b1ecf8019dd53292b7a4d02e16f5/src/templates/partials/logo.html#L25

**Comment by w1ndys** [comment]
Very good tutorial, it increased my knowledge

**Comment by markallasread** [comment]
Excellent thread, thank you for sharing.

**Comment by divSelector** [comment]
Very good approach to finding the template.

**Comment by alexvoss** [comment]
Dear all, please use reactions to say thanks as this reduces the noise for moderators and people who are involved in a thread. I know you mean well but the [contribution guidelines](https://squidfunk.github.io/mkdocs-material/contributing/#before-commenting) ask to make sure to post only things that make a substantive contribution.

**Comment by boogipro** [comment]
This is great thank you!
